File: Sources/App/Services/McpMetadataHealth.swift
import Foundation
import Fluent

/// Dashboard-only MCP metadata quality (mirrors frontend `metadataHealthTier` in `release-skill-metadata-dialog.tsx`).
enum McpMetadataHealth {
    enum Tier: Equatable {
        case blocking
        case warning
        case ok
    }

    /// MCP metadata quality ignoring `compiled.status` — used to infer publish status at compile / PATCH time.
    static func metadataOnlyTier(
        exposureType: String,
        yamlFrontmatterPresent: Bool,
        skillBody: String?,
        schemaJson: String?,
        routing: RoutingHints
    ) -> Tier {
        let raw = schemaJson?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !raw.isEmpty {
            guard (try? JSONSerialization.jsonObject(with: Data(raw.utf8))) != nil else {
                return .blocking
            }
        }
        if exposureType == "resource" {
            if raw.isEmpty {
                return .blocking
            }
            guard let data = raw.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let uri = obj["uri"] as? String,
                  !uri.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                return .blocking
            }
        }
        if !yamlFrontmatterPresent {
            return .warning
        }
        let body = skillBody?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if body.isEmpty {
            return .warning
        }
        if exposureType == "resource" {
            let uw = routing.useWhen ?? []
            let aw = routing.avoidWhen ?? []
            let fm = routing.failureModes ?? []
            if uw.isEmpty && aw.isEmpty && fm.isEmpty {
                return .warning
            }
        }
        return .ok
    }

    /// Combine SKILL inference with metadata tier: blocking → `not_publishable`, warning → `needs_review` unless already not publishable, ok → `inferred`.
    static func resolvedPublishStatus(inferred: String, metadataTier: Tier) -> String {
        switch metadataTier {
        case .blocking:
            return "not_publishable"
        case .warning:
            return inferred == "not_publishable" ? "not_publishable" : "needs_review"
        case .ok:
            return inferred
        }
    }

    static func tier(
        compiled: CompiledSkill,
        schemaJson: String?,
        routing: RoutingHints
    ) -> Tier {
        if compiled.status == "not_publishable" {
            return .blocking
        }
        let meta = metadataOnlyTier(
            exposureType: compiled.exposureType,
            yamlFrontmatterPresent: compiled.yamlFrontmatterPresent,
            skillBody: compiled.skillBody,
            schemaJson: schemaJson,
            routing: routing
        )
        if meta == .blocking {
            return .blocking
        }
        if compiled.status == "needs_review" {
            return .warning
        }
        if meta == .warning {
            return .warning
        }
        if compiled.status == "ready" {
            return .ok
        }
        return .warning
    }

    /// Per-release counts of compiled skills by tier (for release list API).
    static func blockingAndWarningCountsByRelease(
        releaseIds: [UUID],
        compiledSkills: [CompiledSkill],
        schemaBySkillId: [UUID: String],
        ruleBySkillId: [UUID: RoutingRule]
    ) -> [UUID: (blocking: Int, warning: Int)] {
        var blocking: [UUID: Int] = [:]
        var warning: [UUID: Int] = [:]
        for id in releaseIds {
            blocking[id] = 0
            warning[id] = 0
        }
        for cs in compiledSkills {
            let rid = cs.$release.id
            guard let sid = cs.id, releaseIds.contains(rid) else { continue }
            let schema = schemaBySkillId[sid]
            let hints = RoutingHints.from(rule: ruleBySkillId[sid])
            switch tier(compiled: cs, schemaJson: schema, routing: hints) {
            case .blocking:
                blocking[rid] = (blocking[rid] ?? 0) + 1
            case .warning:
                warning[rid] = (warning[rid] ?? 0) + 1
            case .ok:
                break
            }
        }
        var out: [UUID: (blocking: Int, warning: Int)] = [:]
        for id in releaseIds {
            out[id] = (blocking[id] ?? 0, warning[id] ?? 0)
        }
        return out
    }
}

File: services/frontend/lib/mcp-metadata-health.ts
import type { CompiledSkill } from "./types";
import type { McpMetadataFieldId } from "./mcp-metadata-editor-validation";

/** Traffic-light heuristic for MCP metadata quality (dashboard; mirrors backend `McpMetadataHealth.tier`). */
export type McpMetadataHealthTier = "red" | "yellow" | "green";

export function metadataHealthTier(skill: CompiledSkill): McpMetadataHealthTier {
  if (skill.status === "not_publishable") return "red";
  const raw = skill.schema_json?.trim() ?? "";
  if (raw) {
    try {
      JSON.parse(raw);
    } catch {
      return "red";
    }
  }
  if (skill.exposure_type === "resource") {
    if (!raw) return "red";
    try {
      const o = JSON.parse(raw) as { uri?: string };
      if (!o.uri?.trim()) return "red";
    } catch {
      return "red";
    }
  }
  if (skill.status === "needs_review") return "yellow";
  if (skill.yaml_frontmatter_present === false) return "yellow";
  if (!skill.skill_body?.trim()) return "yellow";
  if (skill.exposure_type === "resource") {
    const hasRouting =
      (skill.use_when?.length ?? 0) > 0 ||
      (skill.avoid_when?.length ?? 0) > 0 ||
      (skill.failure_modes?.length ?? 0) > 0;
    if (!hasRouting) return "yellow";
  }
  if (skill.status === "ready") return "green";
  return "yellow";
}

/**
 * For {@link McpMetadataHealthTier} `yellow` skills, which editor fields drive the warning so we can outline them in the MCP metadata UI.
 */
export function mcpReviewHighlightFields(skill: CompiledSkill): McpMetadataFieldId[] {
  if (metadataHealthTier(skill) !== "yellow") return [];

  const fields: McpMetadataFieldId[] = [];
  const add = (f: McpMetadataFieldId) => {
    if (!fields.includes(f)) fields.push(f);
  };

  if (skill.status === "needs_review") add("status");
  if (skill.yaml_frontmatter_present === false) add("skill_body");
  if (!skill.skill_body?.trim()) add("skill_body");
  if (skill.exposure_type === "resource") {
    const hasRouting =
      (skill.use_when?.length ?? 0) > 0 ||
      (skill.avoid_when?.length ?? 0) > 0 ||
      (skill.failure_modes?.length ?? 0) > 0;
    if (!hasRouting) {
      add("use_when");
      add("avoid_when");
      add("failure_modes");
    }
  }
  return fields;
}

/**
 * Best field to scroll to for a skill in the blocking (red) tier.
 * Order matches {@link metadataHealthTier} red checks.
 */
export function mcpFocusFieldForBlockingSkill(skill: CompiledSkill): McpMetadataFieldId {
  if (skill.status === "not_publishable") {
    const sum = skill.summary?.trim() ?? "";
    return sum.length === 0 ? "summary" : "status";
  }
  const raw = skill.schema_json?.trim() ?? "";
  if (raw) {
    try {
      JSON.parse(raw);
    } catch {
      return "schema_json";
    }
  } else if (skill.exposure_type === "resource") {
    return "schema_json";
  }
  if (skill.exposure_type === "resource") {
    try {
      const o = JSON.parse(raw) as { uri?: string };
      if (!o.uri?.trim()) return "schema_json";
    } catch {
      return "schema_json";
    }
  }
  return "schema_json";
}
